src/lib/operators/abstract_operator.hpp

```diff
@@ -50,40 +52,74 @@ enum class OperatorType {
   Mock  // for Tests that need to Mock operators
 };
 
-// AbstractOperator is the abstract super class for all operators.
-// All operators have up to two input tables and one output table.
-// Their lifecycle has three phases:
-// 1. The operator is constructed. Previous operators are not guaranteed to have already executed, so operators must not
-// call get_output in their execute method
-// 2. The execute method is called from the outside (usually by the scheduler). This is where the heavy lifting is done.
-// By now, the input operators have already executed.
-// 3. The consumer (usually another operator) calls get_output. This should be very cheap. It is only guaranteed to
-// succeed if execute was called before. Otherwise, a nullptr or an empty table could be returned.
-//
-// Operators shall not be executed twice.
-//
-// Find more information about operators in our Wiki: https://github.com/hyrise/hyrise/wiki/operator-concept
-
+/**
+ * AbstractOperator is the abstract super class for all operators.
+ * All operators have up to two input tables and one output table.
+ *
+ * LIFECYCLE
+ *  1. The operator is constructed. Because input operators are not guaranteed to have already executed, operators
+ *     must not call get_output in their execute method.
+ *  2. The execute method is called from the outside (usually by the scheduler). This is where the heavy lifting is
+ *     done. By now, the input operators have already executed.
+ *  3. The consumer (usually another operator) calls get_output. This should be very cheap.
+ *
+ *  Operators shall not be executed twice.
+ *
+ * CONSUMER TRACKING
+ *  Operators track the number of consuming operators to automate the clearing of operator results. Therefore,
+ *  an operator registers as a consumer at all of its input operators. After having executed, an operator deregisters
+ *  automatically.
+ *
+ *     WARNING on handling Subqueries:
+ *      This abstract class handles consumer registration/deregistration for input operators only.
+ *      Operators consuming subqueries, such as TableScan and Projection, have to register and deregister as
+ *      consumers manually. For example:
+ *
+ *        1. Projection::Projection
+ *            - Collect uncorrelated subqueries from each expression's arguments.
+ *            - Call register_consumer() and store pointers for all uncorrelated subqueries.
+ *        2. Projection::_on_execute
+ *            - Compute uncorrelated subqueries using ExpressionEvaluator::populate_uncorrelated_subquery_results_cache
+ *            - Call deregister_consumer() for each uncorrelated subquery.
+ *
+ *      It is crucial to call register_consumer() from the constructor, before the execution starts, to prevent subquery
+ *      results from being cleared too early. Otherwise, operators may need to re-execute, which is illegal.
+ *
```

Comment:
Correlated subqueries are deep-copied for each row that they are executed on, so the registration happens at execution time.